sycl/doc/design/OffloadDesign.md

```diff
@@ -0,0 +1,396 @@
+be applied to the `ocloc` step as well.
+
+Depending on the type of binary, the device is embedded as follows:
+
+#### Objects
+
+Object types considered are COFF objects, ELF executables, ELF relocatable
+objects and ELF shared objects.  The device section of the object is in a
+section marked by `.llvm.offloading` for COFF objects.  For ELF files, there is
+a section marked with the `LLVM_OFFLOADING` type.
+
+#### Bitcode
+
+The section which contains the offloading data is from the `.llvm.offloading`
+section within the `llvm.embedded.object` metadata.
+
+#### Offload Binary
+
+The binary itself can be represented by just an offload binary, not requiring
+to be in a section of another binary.  This representation is used for any
+kind of device only binary that is created.
+
+#### Archives
+
+Each item in the archive will be extracted and will be individually checked for
+the file type, properly performing extraction based on the file types already
+listed above.
+
+#### Support of existing fat object format
+
+The current model generates fat objects using the `clang-offload-bundler`
+tool.  This tool creates a binary that contains individual sections that are
+labeled with `__CLANG_OFFLOAD_BUNDLE__kind-triple`.  This format does not
+match the new format as described above.  To enable an ease of transition
+from the old model to the new model, support will be added that allows for
+support of identifying and extracting device binaries that are using the old
+model.
+
+It is expected that all new binaries generated with the updated offloading
+model will represent the embedded fat object format, moving away from the
+`clang-offload-bundler` usage.  We will not support a mixing of fat object
+formats such as objects created with the bundler combined with embedded objects
+in a single archive or objects created with an embedded offload section and
+additional sections generated with the bundler.
+
+### Device Linking
+
+During this phase, all of the individual device binaries that are extracted and
+are associated with a given target are worked on given the type of binary we
+are working with.  The default device code is typically represented in LLVM-IR
+which requires an additional link step of the device code before being
+wrapped and integrated into the final executable.  As mentioned in
+[Packager](#packager) device representation in SPIR-V should be considered
+with the ability to link native SPIR-V modules.  Performing the device link
+against LLVM-IR is performed by `llvm-link`.
+
+To reduce the potential size of the linked device binary, an additional host
+link step is performed to gather dependency information when static device
+libraries are being compiled.  This information is sent through the
+`clang-offload-deps` tool to generate a dependency IR file which is used
+during the device link step.
+
+The use of the clang-linker-wrapper introduces the support of LTO for
```

Comment:
It seems like we need to have 'registered SPIR backend' if we want to use LTO for SPIR targeted IR. How will we accomplish this? Adding @bader here for comments.

Thanks

Reply:
> It seems like we need to have 'registered SPIR backend' if we want to use LTO for SPIR targeted IR. How will we accomplish this?

LLVM has SPIR-V experimental backend (https://github.com/llvm/llvm-project/tree/main/llvm/lib/Target/SPIRV). We can use this backend.
AFAIK, this backend doesn't support all extensions required for DPC++. To mitigate this problem, we can use SPIR-V translator the same ptxas is used for NVPTX target.